Mesh_3/include/CGAL/Mesh_3/mesh_standard_facet_criteria.h

```diff
@@ -626,7 +626,53 @@ class Facet_criterion_visitor
 
 };  // end class Facet_criterion_visitor
 
+template <typename Tr>
+class Facet_criterion_visitor_with_self_intersections
+  : public Mesh_3::Criterion_visitor<Tr, typename Tr::Facet>
+{
+  typedef Mesh_3::Criterion_visitor<Tr, typename Tr::Facet> Base;
+  typedef Facet_criterion_visitor_with_self_intersections<Tr> Self;
+
+public:
+  typedef Mesh_3::Abstract_criterion<Tr, Self> Criterion;
+  typedef typename Base::Quality Facet_quality;
+  typedef typename Base::Is_bad  Is_facet_bad;
+  typedef typename Base::Handle Handle;
+  typedef Handle Facet;
 
+  // Constructor
+  Facet_criterion_visitor_with_self_intersections(const Tr& tr, const Facet& f)
+    : Base(tr, f)
+    , has_self_intersections_(false)
+  {
+    if (f.first->is_facet_on_self_intersection(f.second))
+      has_self_intersections_ = true;
+    else
+    {
+      const Facet& mf = tr.mirror_facet(f);
+      if(mf.first->is_facet_on_self_intersection(mf.second))
+        has_self_intersections_ = true;
```

Comment:
Is that possible that a facet and its mirror facet disagree of the "self intersection" property?

Reply:
they should not

Reply:
Then I think you can just check `f.first->is_facet_on_self_intersection(f.second)`.